### PR TITLE
Add SageMaker support for graph construction, model training and inference.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#59 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/sagemaker/Dockerfile.sm [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+COPY build_artifacts/start_with_right_hostname.sh /usr/local/bin/start_with_right_hostname.sh
+
+# /opt/ml and all subdirectories are utilized by SageMaker, we use the /code subdirectory to store our user code.
+# TODO(xiangsx): change to pip install when PIP package is avaliable
+COPY code/graphstorm/ /opt/ml/code/graphstorm/
+ENV PYTHONPATH="/opt/ml/code/graphstorm/python/:${PYTHONPATH}"
+
+RUN ls /opt/ml/code/graphstorm/
+
+RUN cp /opt/ml/code/graphstorm/python/graphstorm/sagemaker/run/* /opt/ml/code/
```
Comment (thvasilo) [inline docker/sagemaker/Dockerfile.sm:27]: Why are these copied separately?

File: docker/sagemaker/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+```shell
+sudo apt-get update
+sudo apt update
+sudo apt install Docker.io
+```
+
+You need to have the access to pull [SageMaker Framework Contaienrs](https://github.com/aws/deep-learning-containers/blob/master/available_images.md#sagemaker-framework-containers-sm-support-only).
+Please follow [AWS Deep Learning Containers](https://github.com/aws/deep-learning-containers) guideline to get access to the image.
+
+You also need to have the access to [Amazon SageMaker service](https://aws.amazon.com/pm/sagemaker). Getting SageMaker service ready is out of scope of this doc.
```
Comment (thvasilo) [inline docker/sagemaker/README.md:20]:
```suggestion
You also need to have the access to [Amazon SageMaker service](https://aws.amazon.com/pm/sagemaker). To do this you can follow the [SageMaker Getting set-up guide](https://docs.aws.amazon.com/sagemaker/latest/dg/gs-set-up.html).
```

File: docker/sagemaker/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+## Build a Docker image from source
+---------------
+Please use the following command to build a Docker image from source:
+```bash
+mkdir ~/build-docker/
+cd ~/build-docker/
+
+bash /path-to-graphstorm/docker/build_docker_sagemaker.sh /path-to-graphstorm/ docker-name docker-tag
```
Comment (thvasilo) [inline docker/sagemaker/README.md:29]: We can remove `docker-name` and `docker-tag` and the script will still work right?

File: python/graphstorm/sagemaker/README.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
+# Amazon SageMaker Support
+This submodule provides support to run GraphStorm training and inference on Amazon SageMaker.
+To build a SageMaker compatible docker image, please refer to [Build GraphStorm SageMaker docker image] (https://github.com/awslabs/graphstorm/docker/sagemaker).
+
+## Launch SageMaker tasks
+Use scripts under graphstorm.sagemaker.launch to launch SageMaker tasks.
+Please make sure you already setup your SageMaker environments.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:7]:
```suggestion
Please make sure you already setup your SageMaker environment.
```

File: python/graphstorm/sagemaker/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+git clone https://github.com/awslabs/graphstorm.git
+GS_HOME=~/graphstorm/
+python3 $GS_HOME/tools/partition_graph.py --dataset ogbn-arxiv \
+                                          --filepath /tmp/ogbn-arxiv-nc/ \
+                                          --num_parts 2 \
+                                          --output /tmp/ogbn_arxiv_nc_2p
+```
+
+You need to upload /tmp/ogbn_arxiv_nc_2p into S3. You also need to upload the yaml config file into S3.
+You can find the example yaml file in https://github.com/awslabs/graphstorm/blob/main/training_scripts/gsgnn_np/arxiv_nc.yaml.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:26]: Let's put the instructions before the code so the user know what they're running.

File: python/graphstorm/sagemaker/README.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        --task-type "node_classification" \
+        --num-layers 1 \
+        --hidden-size 128 \
+        --backend gloo \
+        --batch-size 128 \
+        --node-feat-name node:feat
+```
+The trained model artifact will be stored in the S3 address provided through `--model-artifact-s3`.
+Please note `save_embed_path` and `save_prediction_path` must be disabled, i.e., set to 'None' when using SageMaker.
+They only work with shared file system while SageMaker solution does not support using shared file system now.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:48]: Same here, instructions before code

File: python/graphstorm/sagemaker/README.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+developing features.
+
+#### Prerequisite 1: Launch or re-use a GPU EC2 instance
+As we will be running multiple heavy containers is one machine we recommend using
+a capable Linux-based machine equiped with GPUs. We recommend at least 32GB of RAM.
+
+
+#### Prerequisite 2: Install Docker and docker compose
+
+You can follow the official Docker guides for [installation of the Docker engine](https://docs.docker.com/engine/install/).
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:118]:
```suggestion
You can follow the official Docker guide for [installation of the Docker engine](https://docs.docker.com/engine/install/).
```

File: python/graphstorm/sagemaker/README.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+Next you need to install the `Docker compose` plugin that will allow us to spin up
+multiple Docker containers. Instructions for that are [here](https://docs.docker.com/compose/install/linux/).
+
+#### Building the SageMaker GraphStorm docker image
+Following [Build GraphStorm SageMaker docker image] (https://github.com/awslabs/graphstorm/docker/sagemaker) to build your own SageMaker GraphStorm docker image.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:124]:
```suggestion
Follow [Build GraphStorm SageMaker docker image] (https://github.com/awslabs/graphstorm/docker/sagemaker) to build your own SageMaker GraphStorm docker image locally.
```

File: python/graphstorm/sagemaker/README.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+```
+Some explanation on the above elements (see the [official docs](https://docs.docker.com/compose/compose-file/) for more details):
+
+* `image`: Determines which image you will use for the container launched.
+* `environment`: Determines the environment variables that will be set for the container once it launches.
+* `command`: Determines the entrypoint, i.e. the command that will be executed once the container launches.
+
+To help you generate yaml file automatically, we provide a Python script that
+builds the docker compose file for your, `generate_sagemaker_docker_compose.py`.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:167]:
```suggestion
builds the docker compose file for you, `generate_sagemaker_docker_compose.py`.
```

File: python/graphstorm/sagemaker/README.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+* `--aws-session-token`: The AWS session toekn used for accessing S3 data within docker.
+
+The rest of the arguments are passed on to `sagemaker_train.py`
+* `--task-type`: Training task type.
+* `--graph-data-s3`: S3 location of input training graph.
+* `--graph-name`: Name of the input training graph.
+* `--train-yaml-s3`: S3 location of training yaml file.
+* `--custom-script`: Custom training script provided by a customer to run customer training logic. This should be a path to the python script within the docker image.
+
+If you want to pass other arguements to `sagemaker_train.py`, you can simply append those arguments after the above arguments.
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:185]:
```suggestion
If you want to pass other arguments to `sagemaker_train.py`, you can simply append those arguments after the above arguments. They will be passed on to the `sagemaker_train.py` script during execution.
```

File: python/graphstorm/sagemaker/README.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+The above will create a Docker compose file named `docker-compose-${task-type}-${num-instances}-train.yml`, which we can then use to launch the job with (for example):
+
+```bash
+docker compose -f docker-compose-node-classification-4-train.yml up
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:190]: Won't the task be using an underscore here? `node_classification`?

File: python/graphstorm/sagemaker/README.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+```
+
+Running the above command will launch 4 instances of the image, configured with
+the command and env vars that emulate a SageMaker execution environment and run
+the `sagemaker_train.py` code. Note that the containers actually
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:195]:
```suggestion
the `sagemaker_train.py` script. Note that the containers actually
```

File: python/graphstorm/sagemaker/README.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+interact with S3 so you would require valid AWS credentials to run.
+
+#### Dcoker compose for inference
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/README.md:198]: Remove? Or add the instructions

File: python/graphstorm/sagemaker/run/sagemaker_infer.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+import sys
+import queue
+
+import boto3
+from graphstorm.config.config import SUPPORTED_TASKS
+from graphstorm.config.config import BUILTIN_TASK_NODE_CLASSIFICATION
+from graphstorm.config.config import BUILTIN_TASK_NODE_REGRESSION
+from graphstorm.config.config import BUILTIN_TASK_EDGE_CLASSIFICATION
+from graphstorm.config.config import BUILTIN_TASK_EDGE_REGRESSION
+from graphstorm.config.config import BUILTIN_TASK_LINK_PREDICTION
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/run/sagemaker_infer.py:37]: Use parentheses for multiple imports ([PEP328](https://peps.python.org/pep-0328/)) e.g.

```
from graphstorm.config.config import (SUPPORTED_TASKS,
    BUILTIN_TASK_NODE_CLASSIFICATION,
    BUILTIN_TASK_NODE_REGRESSION,
    ...
```

Same for all other places multiple imports are done from the same package.

**Review by thvasilo — COMMENTED** [review]
LGTM at the style level, minus the comments. 

We can rely on the CI catching any logic errors, and w.r.t. to the code we'll be making changes, perhaps after the code freeze, to bring up to date to what I have in my own fork.

File: python/graphstorm/sagemaker/README.md (hunk 11 above)
Comment (classicsong) [reply python/graphstorm/sagemaker/README.md:190]: You are right

File: python/graphstorm/sagemaker/README.md (hunk 13 above)
Comment (classicsong) [reply python/graphstorm/sagemaker/README.md:198]: I will add more instructions.

File: python/graphstorm/sagemaker/run/sagemaker_infer.py (hunk 14 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/run/sagemaker_infer.py]: This comment is not resolved right @classicsong ?

**Review by thvasilo — COMMENTED** [review]
LGTM with all the changes. Let's make sure CI passes and merge.

**Pull request merged**